amqp openshiftize

mattf opened the pull request:
in progress set of patches to get caravan running on openshift

currently the generator can publish to a broker, the pathfinder can consume from the broker, the master can consume from pathfinder. the master can run the spark jobs, but the rest notification support is incomplete and the mongodb update is failing with:

```
Traceback (most recent call last):
  File "/opt/spark/python/lib/pyspark.zip/pyspark/streaming/util.py", line 62, in call
    r = self.func(t, *rdds)
  File "/opt/spark/python/lib/pyspark.zip/pyspark/streaming/dstream.py", line 159, in <lambda>
    func = lambda t, rdd: old_func(rdd)
  File "/caravan_master.py", line 69, in <lambda>
    lines.foreachRDD(lambda rdd: process_generic(rdd, mongo_url, rest_url))
  File "/caravan_master.py", line 40, in process_generic
    store_packets(data, mongo_url)
  File "/caravan_master.py", line 23, in store_packets
    db.insert_one(data)
  File "/usr/lib64/python2.7/site-packages/pymongo/collection.py", line 1422, in __call__
    self.__name.split(".")[-1])
TypeError: 'Collection' object is not callable. If you meant to call the 'insert_one' method o
n a 'Collection' object it is failing because no such method exists.
```

Commit:
add master start script to setup name resolution

Commit:
add initial caravan pod definition

Commit:
separated pathfinder from master

the spark streaming receiver is run on a worder,
so it cannot access pathfinder on localhost.
instead, a pathfinder service is created and
used by the master.

Commit:
fix mongodb uri

Commit:
reduce logging

mattf commented:
cc @elmiko

elmiko commented:
hmm, is it possible that machine has pymongo version 2 installed?

i created this with pymongo==3.0.3

also, what is the --mongo string you are passing in?

mattf commented:
the image is centos based and python-pymongo-2.5.2-2.el7.x86_64 is coming from epel.

i can fix that by getting pymongo from pip instlal pymongo=3.0.3 instead

mattf commented:
new error: OperationFailure: not authorized for insert on sparkhara.count_packets

but that might be fixable by adding connection creds to the --mongo url

mattf commented:
hmm, it's a bit of a hole i'm digging here

new error: OperationFailure: command SON([('authenticate', 1), ('user', u'sparkhara'), ('nonce', u'65f26c6ea14716a7'), ('key', u'05a5c5492c72d3c46ce2c00b6efe2b17')]) on namespace admin.$cmd failed: auth fails

mattf commented:
ok, i had a stale image on a worker node. i think the mongo writes are working now because the data in the sparkhara database is increasing.

> db.stats()
> {
>     "db" : "sparkhara",
>     "collections" : 4,
>     "objects" : 55,
>     "avgObjSize" : 1118.3272727272727,
>     "dataSize" : 61508,
>     "storageSize" : 389120,
>     "numExtents" : 4,
>     "indexes" : 3,
>     "indexSize" : 24528,
>     "fileSize" : 16777216,
>     "nsSizeMB" : 16,
>     "dataFileVersion" : {
>         "major" : 4,
>         "minor" : 5
>     },
>     "ok" : 1
> }
> db.stats()
> {
>     "db" : "sparkhara",
>     "collections" : 4,
>     "objects" : 59,
>     "avgObjSize" : 1107.593220338983,
>     "dataSize" : 65348,
>     "storageSize" : 389120,
>     "numExtents" : 4,
>     "indexes" : 3,
>     "indexSize" : 24528,
>     "fileSize" : 16777216,
>     "nsSizeMB" : 16,
>     "dataFileVersion" : {
>         "major" : 4,
>         "minor" : 5
>     },
>     "ok" : 1
> }

Commit:
install pymongo 3.0.3

Commit:
provide a template of a complete connection uri

the username & password will have to change, and
they should be handled by secrets

mattf commented:
@elmiko ptal, no longer wip

elmiko commented on whirlwind_caravan/start_master.sh:
```diff
@@ -0,0 +1,17 @@
+
+spark-submit /caravan_master.py --port 1984 \
+                                --master spark://spark-master:7077 \
+                                --mongo mongodb://sparkhara:openstack@mongodb/sparkhara \
```
only thing to keep in mind here is that the code tries to connect to a database named sparkhara, so i'm not sure how pymongo handles passing the database through the connection string as well.
